package/MDAnalysis/analysis/density.py

```diff
@@ -512,6 +513,10 @@ def _set_user_grid(gridcenter, xdim, ydim, zdim, smin, smax):
         except ValueError as err:
             errmsg = "Non-number values assigned to gridcenter"
             raise ValueError(errmsg) from err
+        if np.isnan(gridcenter).all() and any([xdim, ydim, zdim]):
```

Comment:
It's late in the day so I might be missing something, but the first half of this if is essentially the try condition above right?

If so, it maybe cleaner to make the errmsg value just be conditional of `if any([xdim, ydim, zdim])`?

Reply:
I think the first half is indeed necessary to verify if in the input `gridcenter == None`. In this case the above exception is not raised, but a NumPy array with one `NaN` value is created.
I agree that otherwise it would be cleaner.

Reply:
Then It might be better to put it alongside the if below? Otherwise you end up with an edge case where someone could pass an array of 3 `nan`s (which is possible unintentionally) through and it won't trigger anything?

Reply:
I added the extra check to throw a more informative error to the user in case grid dimensions where provided but no gridcenter. Putting the `any([xdim, ydim, zdim])` alongside the if statement below would reverse this.
If the user inputs grid dimensions my check will catch the edge case, only if the grid dimensions are also missing it wont trigger anything. 
Is in this case catching the edge case more important then an informative error? I assume adding an extra check would be messy.

Reply:
> Is in this case catching the edge case more important then an informative error?
Ideally the answer here is to try to get a solution that covers as much as we can identify. You can't guarantee user behaviour, so when identified edge cases are best covered imho.

My suggestion above was to make the if on the dims a sub-condition of catching either nans or a non-3D shape, but if it's too messy..

> I assume adding an extra check would be messy.

Nothing stops us from re-formatting the way we do the checks 🙃 

How about:

```python
# Move up the `None` check
if any(x is None for x in [gridcenter, xdim, ydim, zdim]):
    # something along the lines that one or more of these are not set
    raise ValueError(errmsg)
# We assume the input should be an ndarray, so the old try/except isn't that useful
if (gridcenter.shape != (3,)
    or np.isnan(gridcenter).any()
    or not np.issubdtype(gridcenter.dtype, np.floating):
        raise ValueError(errmsg)
try:
    xyzdim ....
```

(or some more creative way of doing this)

This has the nice side-effect that if gridcenter is set but the dimensions are not, then you don't accidentally create an ndarray of NaNs.

It introduces that big `if` case in the middle, but it's not exactly really hard to read (at least to me).

That being said, it's late in the day for me, so I'm probably overthinking/being too pedantic. If @fiona-naughton or @hmacdope think it's reasonable as-is, I'd be happy to go forward with it.

Reply:
@IAlibay, I agree, the flow control and error handling is clear here. This layout jams the gridcenter, xdim, ydim, zdim checks in together which is the actual expected behaviour. Thanks for patience @biogen98.

Reply:
> That being said, it's late in the day for me, so I'm probably overthinking/being too pedantic.
@IAlibay I appreciate the suggestions, you raised good points. I will take a look at a nice way to implement it, was just a bit hesitant to change too much of the code I guess.